pythran/pythonic/numpy/around.hpp

```diff
@@ -11,30 +13,42 @@ namespace pythonic
 
   namespace numpy
   {
+    // fast path
+    template <class E>
+    auto around(E const &a) -> decltype(functor::rint{}(a))
+    {
+      return functor::rint{}(a);
+    }
 
     // generic floating point version, pure numpy_expr
     template <class E>
     auto around(E const &a, long decimals) -> typename std::enable_if<
-        !std::is_integral<typename E::dtype>::value,
-        decltype(functor::rint()(a *nt2::pow(typename E::dtype(10), decimals)) /
-                 nt2::pow(typename E::dtype(10), decimals))>::type
+        !std::is_integral<typename types::dtype_of<E>::type>::value,
+        decltype(functor::rint{}(
+                     a *std::declval<typename types::dtype_of<E>::type>()) /
+                 std::declval<typename types::dtype_of<E>::type>())>::type
     {
-      return functor::rint()(a * nt2::pow(typename E::dtype(10), decimals)) /
-             nt2::pow(typename E::dtype(10), decimals);
+      typename types::dtype_of<E>::type const fact = nt2::pow(10., decimals);
+      return functor::rint{}(a * fact) / fact;
     }
 
     // the integer version is only relevant when decimals < 0
     template <class E>
     auto around(E const &a, long decimals) -> typename std::enable_if<
-        std::is_integral<typename E::dtype>::value,
-        decltype(a / (long)nt2::pow(typename E::dtype(10),
-                                    std::max(0L, -decimals)) *
-                 (long)nt2::pow(typename E::dtype(10),
-                                std::max(0L, -decimals)))>::type
+        std::is_integral<typename types::dtype_of<E>::type>::value,
+        decltype((a / std::declval<typename types::dtype_of<E>::type>()) *
+                 std::declval<typename types::dtype_of<E>::type>())>::type
+    {
+      typename types::dtype_of<E>::type const fact =
+          nt2::pow(10L, std::max(0L, -decimals));
+      return (a / fact) * fact;
+    }
+    // list version
+    template <class E>
+    auto around(types::list<E> const &a, long decimals)
+        -> decltype(around(functor::asarray{}(a), decimals))
     {
-      return a /
-             (long)nt2::pow(typename E::dtype(10), std::max(0L, -decimals)) *
-             (long)nt2::pow(typename E::dtype(10), std::max(0L, -decimals));
+      return around(functor::asarray{}(a), decimals);
```

Comment:
We can't do this for now but It could be greate to have an "asarray" function which do nothing for numpy expr so that we can call it in the "main" path